From 9bdda2708efa46a4fc3a26c93acbc61c49f20f0f Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Thu, 12 Mar 2026 18:03:39 +0800
Subject: [PATCH 04/13] feat(imx): add SIP cpu_off service for imx93

SIP cpu_off service is used to turn off specified CPU Core by any
other management CPU Core, SGI interrupt is trigged by managment
CPU Core and handled on specified CPU Core which will be into
WFI in SGI interrupt handler.

Add common layer of inter-core service, so that we can use single
shared SGI interrupt to more inter-core notifyication service.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Upstream-Status: Submitted
---
 plat/imx/common/imx_cpu_off.c               | 43 ++++++++++
 plat/imx/common/imx_intercore_svc.c         | 88 +++++++++++++++++++++
 plat/imx/common/imx_interrupt_mgmt.c        | 11 +--
 plat/imx/common/imx_sip_handler.c           | 27 ++++++-
 plat/imx/common/imx_sip_svc.c               |  4 +
 plat/imx/common/include/imx_cpu_off.h       | 12 +++
 plat/imx/common/include/imx_intercore_svc.h | 42 ++++++++++
 plat/imx/common/include/imx_sip_svc.h       |  6 ++
 plat/imx/imx93/imx93_bl31_setup.c           |  4 +
 plat/imx/imx93/include/platform_def.h       |  3 +
 plat/imx/imx93/platform.mk                  |  7 ++
 11 files changed, 239 insertions(+), 8 deletions(-)
 create mode 100644 plat/imx/common/imx_cpu_off.c
 create mode 100644 plat/imx/common/imx_intercore_svc.c
 create mode 100644 plat/imx/common/include/imx_cpu_off.h
 create mode 100644 plat/imx/common/include/imx_intercore_svc.h

diff --git a/plat/imx/common/imx_cpu_off.c b/plat/imx/common/imx_cpu_off.c
new file mode 100644
index 000000000..1941ac972
--- /dev/null
+++ b/plat/imx/common/imx_cpu_off.c
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <common/debug.h>
+#include <drivers/arm/gicv3.h>
+#include <imx_intercore_svc.h>
+#include <lib/psci/psci.h>
+
+static uint64_t imx_cpu_off(uint32_t id, uint32_t flags,
+		void *handle, void *cookie)
+{
+	unsigned int cpu_id = get_cpu_id();
+	int ret;
+
+	/* Mask all exceptions */
+	write_daifset(0xf);
+	/* Clean and disable SPIs routing to this Core */
+	gicv3_distif_cpu_clear();
+	/* Clean and disable all SGIs/PPIs in this redistributor interface */
+	gicv3_rdistif_cpu_clear(cpu_id);
+	isb();
+	/* Will not return in case of success */
+	ret = psci_cpu_off();
+	if (ret != PSCI_E_SUCCESS) {
+		WARN("CPU off failed\n");
+	}
+
+	return ret;
+}
+
+void imx_cpuoff_init(void)
+{
+	int rc;
+
+	/* CPU off intercore service */
+	rc = register_intercore_cpuoff_svc(imx_cpu_off);
+	if (rc) {
+		panic();
+	}
+}
diff --git a/plat/imx/common/imx_intercore_svc.c b/plat/imx/common/imx_intercore_svc.c
new file mode 100644
index 000000000..78b1d9738
--- /dev/null
+++ b/plat/imx/common/imx_intercore_svc.c
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <common/debug.h>
+#include <drivers/delay_timer.h>
+#include <imx_interrupt_mgmt.h>
+#include <imx_intercore_svc.h>
+#include <imx_interrupt_mgmt.h>
+#include <plat/common/platform.h>
+#include <platform_def.h>
+
+static percpu_svc_inf_t svc_info[PLATFORM_CORE_COUNT];
+
+static uint64_t logical_id_to_hwid(unsigned int cpu_id)
+{
+	uint64_t hwid = cpu_id;
+
+	if ((read_mpidr_el1() & MPIDR_MT_MASK) != 0U)
+		hwid = cpu_id << MPIDR_AFF1_SHIFT;
+
+	return hwid;
+}
+
+int register_intercore_cpuoff_svc(interrupt_type_handler_t handler)
+{
+	percpu_svc_inf_t *inf;
+	uint32_t i;
+
+	if (!handler) {
+		return -EINVAL;
+	}
+
+	for (i = 0; i < PLATFORM_CORE_COUNT; i++) {
+		inf = &svc_info[i];
+		inf->cpu_off_svc.enabled = true;
+		inf->cpu_off_svc.handler = handler;
+	}
+
+	return 0;
+}
+
+int trigger_intercore_cpuoff_svc(unsigned int cpu)
+{
+	percpu_svc_inf_t *inf;
+
+	if (cpu >= PLATFORM_CORE_COUNT) {
+		return -EINVAL;
+	}
+
+	inf = &svc_info[cpu];
+	inf->cpu_off_svc.trigged = true;
+	inf->cpu_off_svc.result = 0;
+
+	plat_ic_raise_el3_sgi(INTERCORE_SVC_IRQ, logical_id_to_hwid(cpu));
+
+	/* Wait 1ms in case of returned error from cpu_off_svc handler */
+	mdelay(1);
+
+	return inf->cpu_off_svc.result;
+}
+
+static uint64_t intercore_service(uint32_t id, uint32_t flags,
+		void *handle, void *cookie)
+{
+	unsigned int cpu_id = get_cpu_id();
+	percpu_svc_inf_t *inf = &svc_info[cpu_id];
+
+	if (inf->cpu_off_svc.enabled && inf->cpu_off_svc.trigged && inf->cpu_off_svc.handler) {
+		inf->cpu_off_svc.trigged = false;
+		inf->cpu_off_svc.result = inf->cpu_off_svc.handler(id, flags, handle, cookie);
+	}
+
+	return 0;
+}
+
+void imx_intercore_svc_init(void)
+{
+	int rc;
+
+	/* Register the EL3 handler for Intercore service */
+	rc = request_intr_type_el3(INTERCORE_SVC_IRQ, intercore_service);
+	if (rc) {
+		panic();
+	}
+}
diff --git a/plat/imx/common/imx_interrupt_mgmt.c b/plat/imx/common/imx_interrupt_mgmt.c
index 33c1a837f..0e2fe5f90 100644
--- a/plat/imx/common/imx_interrupt_mgmt.c
+++ b/plat/imx/common/imx_interrupt_mgmt.c
@@ -41,20 +41,17 @@ static uint64_t imx_el3_interrupt_handler(uint32_t id, uint32_t flags,
 		WARN("No pending interrupt\n");
 		return 0U;
 	}
-	intr_id = plat_ic_acknowledge_interrupt();
 
-	INFO("Interrupt recvd is %d\n", intr_id);
+	intr_id = plat_ic_acknowledge_interrupt();
+	/* Mark this interrupt as complete firstly as CPU Off will not return from handler */
+	plat_ic_end_of_interrupt(intr_id);
+	isb();
 
 	handler = type_el3_interrupt_table[intr_id];
 	if (handler != NULL) {
 		handler(intr_id, flags, handle, cookie);
 	}
 
-	/*
-	 * Mark this interrupt as complete to avoid a interrupt storm.
-	 */
-	plat_ic_end_of_interrupt(intr_id);
-
 	return 0U;
 }
 
diff --git a/plat/imx/common/imx_sip_handler.c b/plat/imx/common/imx_sip_handler.c
index ddcd3269e..21e64eca0 100644
--- a/plat/imx/common/imx_sip_handler.c
+++ b/plat/imx/common/imx_sip_handler.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 NXP
+ * Copyright 2019,2024 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -13,8 +13,10 @@
 #include <common/debug.h>
 #include <common/runtime_svc.h>
 #include <platform_def.h>
+#include <imx_intercore_svc.h>
 #include <imx_sip_svc.h>
 #include <lib/el3_runtime/context_mgmt.h>
+#include <lib/el3_runtime/cpu_data.h>
 #include <lib/mmio.h>
 #include <sci/sci.h>
 #include <drivers/arm/gic_common.h>
@@ -389,3 +391,26 @@ int imx_bbsm_handler(uint32_t smc_fid, u_register_t x1, void *handle)
 	return ret;
 }
 #endif
+
+#if defined(PLAT_imx93)
+uint64_t imx_sip_cpuoff_handler(uint32_t smc_fid,
+			u_register_t x1,
+			u_register_t x2,
+			u_register_t x3,
+			u_register_t x4)
+{
+	uint32_t cpu_id = x1;
+	int ret;
+
+	if (get_cpu_data_by_index(cpu_id, psci_svc_cpu_data.aff_info_state) == AFF_STATE_ON) {
+		ret = trigger_intercore_cpuoff_svc(cpu_id);
+		if (ret) {
+			return SMC_UNK;
+		}
+
+		return SMC_OK;
+	}
+
+	return SMC_UNK;
+}
+#endif
diff --git a/plat/imx/common/imx_sip_svc.c b/plat/imx/common/imx_sip_svc.c
index d177caf21..55dcf8c71 100644
--- a/plat/imx/common/imx_sip_svc.c
+++ b/plat/imx/common/imx_sip_svc.c
@@ -99,6 +99,10 @@ static uintptr_t imx_sip_handler(unsigned int smc_fid,
 	case IMX_SIP_GET_SOC_INFO:
 		return imx9_soc_info_handler(smc_fid, handle);
 #endif
+#if defined(PLAT_imx93)
+	case IMX_SIP_CPU_OFF:
+		SMC_RET1(handle, imx_sip_cpuoff_handler(smc_fid, x1, x2, x3, x4));
+#endif
 #if defined(PLAT_imx93) || defined(PLAT_imx91)
 	case IMX_SIP_DDR_DVFS:
 		return dram_dvfs_handler(smc_fid, handle, x1, x2, x3);
diff --git a/plat/imx/common/include/imx_cpu_off.h b/plat/imx/common/include/imx_cpu_off.h
new file mode 100644
index 000000000..20d27c9c0
--- /dev/null
+++ b/plat/imx/common/include/imx_cpu_off.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef IMX_CPU_OFF_H
+#define IMX_CPU_OFF_H
+
+void imx_cpuoff_init(void);
+
+#endif		/*	IMX_CPU_OFF_H	*/
diff --git a/plat/imx/common/include/imx_intercore_svc.h b/plat/imx/common/include/imx_intercore_svc.h
new file mode 100644
index 000000000..316f34da4
--- /dev/null
+++ b/plat/imx/common/include/imx_intercore_svc.h
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef IMX_INTERCORE_SVC_H
+#define IMX_INTERCORE_SVC_H
+
+#include <arch_helpers.h>
+
+#include <platform_def.h>	/* CACHE_WRITEBACK_GRANULE required */
+#include <stdbool.h>
+#include <stdint.h>
+#include <imx_interrupt_mgmt.h>
+
+typedef struct intercore_svc {
+	volatile bool enabled;
+	volatile bool trigged;
+	volatile int result;
+	interrupt_type_handler_t handler;
+} intercore_svc_t;
+
+typedef struct percpu_svc_inf {
+	intercore_svc_t cpu_off_svc;
+} __aligned(CACHE_WRITEBACK_GRANULE) percpu_svc_inf_t;
+
+static inline unsigned int get_cpu_id(void)
+{
+	uint64_t mpidr = read_mpidr_el1();
+
+	if ((mpidr & MPIDR_MT_MASK) != 0U)
+		return MPIDR_AFFLVL1_VAL(mpidr);
+	else
+		return MPIDR_AFFLVL0_VAL(mpidr);
+}
+
+int register_intercore_cpuoff_svc(interrupt_type_handler_t handler);
+int trigger_intercore_cpuoff_svc(unsigned int cpu);
+void imx_intercore_svc_init(void);
+
+#endif	/* IMX_INTERCORE_SVC_H	*/
diff --git a/plat/imx/common/include/imx_sip_svc.h b/plat/imx/common/include/imx_sip_svc.h
index 2314b87e1..7564a9355 100644
--- a/plat/imx/common/include/imx_sip_svc.h
+++ b/plat/imx/common/include/imx_sip_svc.h
@@ -75,6 +75,8 @@
 #define IMX_SIP_BBSM_READ_TAMPER_STATUS 0x02
 #endif
 
+#define IMX_SIP_CPU_OFF			0xC2000012
+
 #define IMX_SIP_AARCH32			0xC20000FD
 
 int imx_kernel_entry_handler(uint32_t smc_fid, u_register_t x1,
@@ -177,4 +179,8 @@ int imx_src_handler(uint32_t smc_fid, u_register_t x1,
 		    u_register_t x2, u_register_t x3,
 		    u_register_t x4, void *handle);
 #endif
+#if defined(PLAT_imx93)
+uint64_t imx_sip_cpuoff_handler(uint32_t smc_fid, u_register_t x1,
+	u_register_t x2, u_register_t x3, u_register_t x4);
+#endif
 #endif /* __IMX_SIP_SVC_H__ */
diff --git a/plat/imx/imx93/imx93_bl31_setup.c b/plat/imx/imx93/imx93_bl31_setup.c
index 954d7b608..ff9ea771f 100644
--- a/plat/imx/imx93/imx93_bl31_setup.c
+++ b/plat/imx/imx93/imx93_bl31_setup.c
@@ -21,6 +21,8 @@
 
 #include <ele_api.h>
 #include <dram.h>
+#include <imx_cpu_off.h>
+#include <imx_intercore_svc.h>
 #include <imx_interrupt_mgmt.h>
 #include <imx8_lpuart.h>
 #include <plat_common.h>
@@ -159,6 +161,8 @@ void bl31_platform_setup(void)
 	ele_get_soc_info();
 
 	imx_el3_interrupt_config();
+	imx_intercore_svc_init();
+	imx_cpuoff_init();
 
 	/* Init the dram info */
 	dram_info_init(SAVED_DRAM_TIMING_BASE);
diff --git a/plat/imx/imx93/include/platform_def.h b/plat/imx/imx93/include/platform_def.h
index f590f4258..b990b4af2 100644
--- a/plat/imx/imx93/include/platform_def.h
+++ b/plat/imx/imx93/include/platform_def.h
@@ -129,4 +129,7 @@
 #define IMX_TRUSTY_STACK_SIZE 0x200
 #define TRUSTY_SHARED_MEMORY_OBJ_SIZE (12 * 1024)
 
+/*Use SGI-8 as Inter-Core Service IRQ */
+#define INTERCORE_SVC_IRQ		0x8
+
 #endif /* platform_def.h */
diff --git a/plat/imx/imx93/platform.mk b/plat/imx/imx93/platform.mk
index 8c9a7bbd9..bfa231f5a 100644
--- a/plat/imx/imx93/platform.mk
+++ b/plat/imx/imx93/platform.mk
@@ -31,7 +31,11 @@ BL31_SOURCES		+=	common/desc_image_load.c			\
 				plat/imx/imx93/aarch64/plat_helpers.S		\
 				plat/imx/imx93/plat_topology.c			\
 				plat/imx/common/imx_common.c			\
+				plat/imx/common/imx_cpu_off.c			\
+				plat/imx/common/imx_intercore_svc.c		\
 				plat/imx/common/imx_interrupt_mgmt.c		\
+				plat/imx/common/imx_sip_handler.c		\
+				plat/imx/common/imx_sip_svc.c			\
 				plat/imx/common/lpuart_console.S		\
 				plat/imx/imx93/trdc.c			\
 				plat/imx/imx93/pwr_ctrl.c			\
@@ -59,6 +63,9 @@ USE_COHERENT_MEM	:=	0
 PROGRAMMABLE_RESET_ADDRESS :=	1
 COLD_BOOT_SINGLE_CPU	:=	1
 
+# Enable workaround for irq losing in case of pending NS SError
+HANDLE_NS_EA_EL3_IRQ_WA	:=	1
+
 BL32_BASE               ?=      0x96000000
 BL32_SIZE               ?=      0x02000000
 $(eval $(call add_define,BL32_BASE))
-- 
2.43.0

